poller/poller.go

```diff
@@ -4,39 +4,69 @@ import (
 	"time"
 
 	"github.com/godaddy/split-go-serializer/api"
+	"github.com/splitio/go-client/splitio/service/dtos"
 )
 
-// Poller contains cache data and SplitioAPIBinding
+// Poller contains cache data, splitioDataGetter, and required info to interact with aplitio api
 type Poller struct {
-	Cache              int
-	Error              error
-	splitioAPIBinding  api.SplitioAPIBinding
+	Cache              Cache
+	Error              chan error
+	splitioDataGetter  api.SplitioDataGetter
 	pollingRateSeconds int
 	serializeSegments  bool
 	quit               chan bool
 }
 
+// Cache contains Splits and Segments which is supposed to be updated periodically
+type Cache struct {
+	Splits             []dtos.SplitDTO
+	Since              int64
+	Segments           []dtos.SegmentChangesDTO
+	UsingSegmentsCount int
+}
+
 // NewPoller returns a new Poller
-func NewPoller(splitioAPIKey string, pollingRateSeconds int, serializeSegments bool) *Poller {
+func NewPoller(splitioAPIKey string, pollingRateSeconds int, serializeSegments bool, splitioDataGetter api.SplitioDataGetter) *Poller {
 	if pollingRateSeconds == 0 {
 		pollingRateSeconds = 300
 	}
+	if splitioDataGetter != nil {
+		return &Poller{Cache{}, make(chan error), splitioDataGetter, pollingRateSeconds, serializeSegments, make(chan bool)}
+	}
 	splitioAPIBinding := api.NewSplitioAPIBinding(splitioAPIKey, "")
-	return &Poller{0, nil, *splitioAPIBinding, pollingRateSeconds, serializeSegments, make(chan bool)}
+	return &Poller{Cache{}, make(chan error), splitioAPIBinding, pollingRateSeconds, serializeSegments, make(chan bool)}
 }
 
-// pollForChanges will get the latest data of splits and segment
+// pollForChanges updates the Cache with latest splits and segment
 func (poller *Poller) pollForChanges() {
-	// TODO: call getSplits and getSegments to formulate the cache
-	// if any of the returned splits/segments have error:
-	// 1. pass the error to poller.Error and log the error
-	// 2. if cache updates successfully, set poller.Error to nil
-	poller.Cache++
+	binding := poller.splitioDataGetter
+	splits, since, err := binding.GetSplits()
+	if err != nil {
+		poller.Error <- err
+		return
+	}
+
+	segments := []dtos.SegmentChangesDTO{}
+	usingSegmentsCount := 0
+	if poller.serializeSegments {
+		segments, usingSegmentsCount, err = binding.GetSegmentsForSplits(splits)
+		if err != nil {
+			poller.Error <- err
+			return
+		}
+	}
+
+	poller.Cache = Cache{
```

Comment:
this assignment isn't going to atomically set all values in `poller.Cache`. that makes it impossible to read a consistent value for `poller.Cache` (you might get a mix of old and new values).

Use `atomic.LoadPointer` and `atomic.StorePointer`:

```go
type Poller struct {
  ...
  cache unsafe.Pointer
}

// Update cache
atomic.StorePointer(&poller.cache, unsafePointer(&Cache{ ... }))

// Return cache resutls
func (poller *Poller) GetCache() Cache {
  return *(*Cache)(atomic.LoadPointer(&poller.cache)))
}
```